kubeadm: use the new v1beta3 instead of v1beta2 (#28360)
kubernetes/website: content/en/docs/setup/production-environment/tools/kubeadm/control-plane-flags.md

For more details on each field in the configuration you can navigate to our
[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta3).

{{< note >}}
Customizing the CoreDNS deployment of kubeadm is currently not supported. You must manually
patch the `kube-system/coredns` {{< glossary_tooltip text="ConfigMap" term_id="configmap" >}}
and recreate the CoreDNS {{< glossary_tooltip text="Pods" term_id="pod" >}} after that. Alternatively,
you can skip the default CoreDNS deployment and deploy your own variant.
For more details on that see [Using init phases with kubeadm](/docs/reference/setup-tools/kubeadm/kubeadm-init/#init-phases).
{{< /note >}}

<!-- body -->

{{< feature-state for_k8s_version="v1.12" state="stable" >}}

## Customizing the control plane with flags in `ClusterConfiguration`

The kubeadm `ClusterConfiguration` object exposes the field `extraArgs` that can override the default
flags passed to control plane components such as the APIServer, ControllerManager, Scheduler and Etcd.
The components are defined using the following fields:

- `apiServer`
- `controllerManager`
- `scheduler`
- `etcd`

> The `extraArgs` field consist of `key: value` pairs. To override a flag for a control plane component:

Reviewer: s/consist/consists ?

small nit

1.  Add the appropriate fields to your configuration.

> 2.  Add the flags to override to the field.

Reviewer: `to override to the field`
to
`to override the field`

Author: reworded this block and list to clarify the concepts of structures / field.

3.  Run `kubeadm init` with `--config <YOUR CONFIG YAML>`.

{{< note >}}
You can generate a `ClusterConfiguration` object with default values by running `kubeadm config print init-defaults`
and saving the output to a file of your choice.
{{< /note >}}

{{< note >}}
The `ClusterConfiguration` object is currently global in kubeadm clusters. This means that any flags that you add,
will apply to all instances of the same component on different nodes. To apply individual configuration per component
on different nodes you can use [patches](#patches).
{{< /note >}}

{{< note >}}

> Duplicate flags (keys), or passing the same flag `--foo` multiple times, is currently not supported.

Reviewer: ```
or passing the same flag `--foo` multiple times
```
Does this dup with `duplicate flags(keys)`? 🤔

Author: it's just a clarification for the first part of the sentence.

{{< note >}}
Customizing the CoreDNS deployment of kubeadm is currently not supported. You must manually
patch the `kube-system/coredns` {{< glossary_tooltip text="ConfigMap" term_id="configmap" >}}
and recreate the CoreDNS {{< glossary_tooltip text="Pods" term_id="pod" >}} after that. Alternatively,
you can skip the default CoreDNS deployment and deploy your own variant.
For more details on that see [Using init phases with kubeadm](/docs/reference/setup-tools/kubeadm/kubeadm-init/#init-phases).
{{< /note >}}

<!-- body -->

{{< feature-state for_k8s_version="v1.12" state="stable" >}}

## Customizing the control plane with flags in `ClusterConfiguration`

The kubeadm `ClusterConfiguration` object exposes a way for users to override the default
flags passed to control plane components such as the APIServer, ControllerManager, Scheduler and Etcd.
The components are defined using the following structures:

- `apiServer`
- `controllerManager`
- `scheduler`
- `etcd`

These structures contain a common `extraArgs` field, that consists of `key: value` pairs.
To override a flag for a control plane component:

1.  Add the appropriate `extraArgs` to your configuration.

> 2.  Add flags to the `extraArgs` fields.

Reviewer 2: ```suggestion
2.  Add flags to the `extraArgs` field.
```

      pathType: "File"
```

### Etcd flags

For details, see the [etcd server documentation](https://etcd.io/docs/).

Example usage:

```yaml
apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
etcd:
  local:
    extraArgs:
      election-timeout: 1000
```

## Customizing the control plane with patches {#patches}

{{< feature-state for_k8s_version="v1.22" state="beta" >}}

> Kubeadm allows you pass a directory with patch files to `InitConfiguration` and `JoinConfiguration`

Reviewer 2: ```suggestion
Kubeadm allows you to pass a directory with patch files to `InitConfiguration` and `JoinConfiguration`
```

on individual nodes. These patches can be used as the last customization step before the control

> plane component manifests are written on disk.

Reviewer 2: ```suggestion
plane component manifests are written to disk.
```

You can pass this file to `kubeadm init` with `--config <YOUR CONFIG YAML>`:

```yaml
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration
nodeRegistration:
  patches:
    directory: /home/user/somedir
```

{{< note >}}
For `kubeadm init` you can pass a file containing both a `ClusterConfiguration` and `InitConfiguration`
separated by `---`.
{{< /note >}}

You can pass this file to `kubeadm join` with `--config <YOUR CONFIG YAML>`:

```yaml
apiVersion: kubeadm.k8s.io/v1beta3
kind: JoinConfiguration
nodeRegistration:
  patches:
    directory: /home/user/somedir
```

The directory must contain files named `target[suffix][+patchtype].extension`.
For example, `kube-apiserver0+merge.yaml` or just `etcd.json`.

- `target` can be one of `kube-apiserver`, `kube-controller-manager`, `kube-scheduler` and `etcd`.

> - `patchtype` can be one of `strategic`, `merge` or `json` and these match the patch formats

Reviewer 2: ```suggestion
- `patchtype` can be one of `strategic`, `merge` or `json` and these must match the patching formats
```

[supported by kubectl](/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch).
The default `patchtype` is `strategic`.
- `extension` must be either `json` or `yaml`.
- `suffix` is an optional string that can be used to determine which patches are applied first
alpha-numerically.

{{< note >}}
If you are using `kubeadm upgrade` to upgrade your kubeadm nodes you must again provide the same
patches, so that the customization is preserved after upgrade. To do that you can use the `--patches`
flag, which must point to the same directory. `kubeadm upgrade` currently does not support a configuration
API structure that can be used for the same purpose.
{{< /note >}}

## Customizing the kubelet

> To customize the kubelet you can pass a `KubeletConfiguration` next your `ClusterConfiguration` or

Reviewer 2: ```suggestion
To customize the kubelet you can pass a `KubeletConfiguration` next to the `ClusterConfiguration` or
```

Author: reworded a little the two adjacent sentences.
